File: src/main/java/com/projet/korector/services/SessionService.java
package com.projet.korector.services;

import com.projet.korector.controller.SessionController;
import com.projet.korector.controller.UserController;
import com.projet.korector.entity.*;
import com.projet.korector.repository.*;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import javax.servlet.http.HttpServletResponse;
import java.io.Serializable;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class SessionService {

    final static Logger log = LoggerFactory.getLogger(SessionController.class);

    @Autowired
    private SessionRepository sessionRepository;

    @Autowired
    private RunRepository runRepository;

    @Autowired
    private ProjectRepository projectRepository;

    @Autowired
    private CriteriaRepository criteriaRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private UserController userController;

    @Autowired
    private SessionCritereRepository sessionCritereRepository;

    public ResponseEntity<Session> createSession(SessionImp sessionImp, User currentUser)
    {
        log.info("objet angular reçu pour création :"+sessionImp.toString());

        Session session = new Session(sessionImp.getName(),sessionImp.getDate_depot(),sessionImp.getHeureDepot());

        Set<Project> projects = new HashSet<>();
        sessionImp.getProjects().forEach(projectId-> {
            projects.add(this.projectRepository.findById(projectId).get());
        });

        session.setProjects(new HashSet<>());
        Session createdSession = this.sessionRepository.save(session);

        createdSession.getProjects().addAll(projects);

        log.info("createdSession : "+createdSession);
        currentUser.getSessions().add(createdSession);
        createdSession.getUsers().add(currentUser);

        return new ResponseEntity<Session>(this.sessionRepository.saveAndFlush(createdSession), HttpStatus.OK);
    }

    public Session updateSession(SessionImp sessionImp)
    {
        log.info("Session à modifier : "+sessionImp);
        Session session=new Session(sessionImp.getId(),sessionImp.getName(),sessionImp.getDate_depot(),sessionImp.getHeureDepot());
        sessionImp.getProjects().forEach(projectId-> {
            session.getProjects().add(this.projectRepository.findById(projectId).get());
        });

        sessionImp.getSessionCritere().forEach(sessionCritereId-> {
            session.getSessionCriteres().add(this.sessionCritereRepository.findById(sessionCritereId).get());
        });

        Session newSession=null;
        if(this.sessionRepository.findById(session.getId()).isPresent())
        {
            this.setSessionProjects(session.getId(),session.getProjects());
            this.setSessionCriteres(session.getId(),session.getSessionCriteres());

            newSession=this.sessionRepository.save(session);
        }
        log.info("UpdateSession : "+newSession);
        return newSession;
    }

    public List<Session> getAllSessions()
    {
        return sessionRepository.findAll();
    }

    public Set<Session> getAllSessionsByUser(User currentUser) {

        return sessionRepository.findAll().stream().filter(session -> session.getUsers().contains(currentUser)).collect(Collectors.toSet());
    }

    public Set<Session> getSessionWithDateDepotNotNull()
    {
        Set<Session> sessions = this.sessionRepository.findAll().stream().filter(session -> !session.getDate_depot().equals("")).collect(Collectors.toSet());
        log.info("Session- date depot not null : "+sessions);
        return sessions;
    }

    public Set<Project> getSessionProjects(Long sessionId)
    {
        return new HashSet<Project>(this.sessionRepository.findById(sessionId).get().getProjects());
    }

//    public Set<Criteria> getSessionCriterias(Long sessionId)
//    {
//        return new HashSet<Criteria>(this.sessionRepository.findById(sessionId).get().getCriterias());
//    }

    public Set<SessionCritere> getSessionCriteres(Long sessionId)
    {
        log.info("SessionCriteres : "+this.sessionRepository.findById(sessionId).get().getSessionCriteres());
        return  new HashSet<>(this.sessionRepository.findById(sessionId).get().getSessionCriteres());
    }

    public void addProjectToSession(Long sessionId,Long projectId)
    {
        log.info(" ************** Add Project : "+projectId+" to session : "+sessionId);
        System.out.println(" ************** Add Project : "+projectId+" to session : "+sessionId);
        Project putProject=this.projectRepository.findById(projectId).get();
        Session putSession=this.sessionRepository.findById(sessionId).get();

        putSession.getProjects().add(putProject);
        putProject.getSessions().add(putSession);

        this.projectRepository.saveAndFlush(putProject);
        this.sessionRepository.saveAndFlush(putSession);
    }

    public void setSessionProjects(Long sessionId,Set<Project> projects)
    {
        Set<Project> putProjects= new HashSet<>();
        projects.forEach(project->{
            putProjects.add(this.projectRepository.findById(project.getId()).get());
        });

        Session putSession=this.sessionRepository.findById(sessionId).get();
        putSession.setProjects(putProjects);

        putProjects.forEach(project -> {
            project.getSessions().add(putSession);
            this.projectRepository.saveAndFlush(project);
        });

        this.sessionRepository.saveAndFlush(putSession);
    }

//    public void setSessionCriterias(Long sessionId,Set<Criteria> criterias)
//    {
//        Set<Criteria> putCriterias= new HashSet<>();
//        criterias.forEach(criteria->{
//            putCriterias.add(this.criteriaRepository.findById(criteria.getId()).get());
//        });
//
//        Session putSession=this.sessionRepository.findById(sessionId).get();
//        putSession.setCriterias(putCriterias);
//
//        putCriterias.forEach(criteria -> {
//            criteria.getSessions().add(putSession);
//            this.criteriaRepository.saveAndFlush(criteria);
//        });
//
//        this.sessionRepository.saveAndFlush(putSession);
//    }

    public void setSessionCriteres(Long sessionId,Set<SessionCritere> sessionCriteres)
    {
        Set<SessionCritere> putSessionCriteres= new HashSet<>();
        sessionCriteres.forEach(sessionCritere->{
            putSessionCriteres.add(this.sessionCritereRepository.findById(sessionCritere.getId()).get());
        });

        Session putSession=this.sessionRepository.findById(sessionId).get();
        putSession.setSessionCriteres(putSessionCriteres);

        putSessionCriteres.forEach(sessionCritere -> {
            sessionCritere.setSession(putSession);
            this.sessionCritereRepository.saveAndFlush(sessionCritere);
        });

        this.sessionRepository.saveAndFlush(putSession);
    }

    public void deleteProjectFromSession(Long sessionId,Long projectId)
    {
        Project deleteProject=this.projectRepository.findById(projectId).get();
        Session putSession=this.sessionRepository.findById(sessionId).get();

        putSession.getProjects().remove(deleteProject);
        deleteProject.getSessions().remove(putSession);
        this.projectRepository.saveAndFlush(deleteProject);
        this.sessionRepository.saveAndFlush(putSession);
    }

    public void deleteSession(Long id, User user){
        Session deletedSession=this.sessionRepository.findById(id).get();
        boolean isPresent = this.sessionRepository.findById(id).isPresent();
        if(isPresent)
        {
            Set<Project> projects= deletedSession.getProjects();
            Set<SessionCritere> sessionCriteres = deletedSession.getSessionCriteres();
            Set<Run> runs = deletedSession.getRuns();

            projects.forEach(project -> {
                this.projectRepository.findById(project.getId()).get().getSessions().remove(deletedSession);
                this.projectRepository.saveAndFlush(project);
            });

            sessionCriteres.forEach(sessionCritere -> {
                this.sessionCritereRepository.deleteById(sessionCritere.getId());
            });

            projects.clear();
            sessionCriteres.clear();
            runs.clear();

            user.getSessions().remove(deletedSession);
            this.userRepository.save(user);
            deletedSession.getUsers().remove(user);

            this.sessionRepository.saveAndFlush(deletedSession);
            this.sessionRepository.deleteById(id);
        }
    }

    public void deleteAllSessions(User user)
    {
        this.sessionRepository.findAll().forEach(session -> {
            this.deleteSession(session.getId(), user);
        });
    }

    public Session getSessionById(Long sessionId) {
        return this.sessionRepository.findById(sessionId).orElse(null);
    }

    public Set<Run> getSessionRuns(Long sessionId)
    {
        Set<Run> runs = new HashSet<Run>();
        runs.addAll(this.sessionRepository.findById(sessionId).get().getRuns());
        return runs;
    }



    public void exportCSV(Long runId, HttpServletResponse response) {

        SimpleDateFormat formater = new SimpleDateFormat("yyyyMMddHHmmss");
        Date date = new Date();

        //set file name and content type
        String filename = "run_"+date+".csv";

        response.setContentType("text/csv");
        response.setHeader(HttpHeaders.CONTENT_DISPOSITION,
                "attachment; filename=\"" + filename + "\"");

        //create a csv writer
        Run run = this.runRepository.findById(runId).get();

        List<Project> sessionsProject = new ArrayList<Project>(this.getSessionProjects(run.getSession().getId()));

        try (
                CSVPrinter csvPrinter = new CSVPrinter(response.getWriter(), CSVFormat.DEFAULT
                        .withHeader("ID", "FirstName", "LastName"));
        ) {
            for (Project project : sessionsProject) {
                List<? extends Serializable> data = Arrays.asList(
                        project.getId(),
                        project.getName(),
                        project.getNote(),
                        project.getUrl()
                );
                csvPrinter.printRecord(data);
            }
            csvPrinter.flush();
        } catch (Exception e) {
            System.out.println("Writing CSV error!");
            e.printStackTrace();
        }
    }
}

File: src/main/java/com/projet/korector/frontend/src/app/_services/session.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { environment } from 'src/environments/environment';
import { Session } from '../classes/session';
import { Project } from '../classes/project';
import { Run } from '../classes/run';
import { Criteria } from '../classes/criteria';
import { SessionCritere } from '../classes/session-critere';

const httpOptions = {
  headers: new HttpHeaders(
  {
     'Content-Type': 'application/json'
  })
}

@Injectable({
  providedIn: 'root'
})

export class SessionService {

  public url =environment.api_url;
  constructor(private http: HttpClient) {
  }

  public createSession(session : Session) : Observable<Session>
  {
    const routeQuery=this.url+"/createSession";
    console.log("session", session)

    console.log("session dans service : ", session)
    return this.http.post<Session>(routeQuery,session);
  }

  public updateSession(session: Session) : Observable<any>
  {
    const routeQuery=this.url+"/updateSession";
    console.log("updateSession", session)

    return this.http.put<Session>(routeQuery,session);
  }

  public getSessionById(sessionId : Number) : Observable<Session>
  {
    const routeQuery=this.url+"/sessionById/"+sessionId;
    return this.http.get<Session>(routeQuery);
  }

  public getAllSessions() : Observable<Array<Session>>
  {
    const routeQuery=this.url+"/allSessions";
    return this.http.get<Array<Session>>(routeQuery);
  }

  public getAllSessionsByUser() : Observable<Array<Session>>
  {
    const routeQuery=this.url+"/user/sessions";
    return this.http.get<Array<Session>>(routeQuery);
  }

  public getSessionProjects(sessionId : Number) : Observable<Array<Project>>
  {
    let routeQuery=this.url+"/sessionProjects/"+sessionId;
    return this.http.get<Array<Project>>(routeQuery);
  }

  public getSessionCriterias(sessionId : Number) : Observable<Array<Criteria>>
  {
    let routeQuery=this.url+"/sessionCriterias/"+sessionId;
    return this.http.get<Array<Criteria>>(routeQuery);
  }

  public getSessionCriteres(sessionId : number) : Observable<Array<SessionCritere>>
  {
    let routeQuery=this.url+"/getSessionCriteres/"+sessionId;
    return this.http.get<Array<SessionCritere>>(routeQuery);
  }

  public deleteSession (sessionId :Number) : Observable<any>
  {
    const routeQuery=this.url+"/deleteSession/"+sessionId;
    return this.http.delete(routeQuery);
  }

  public deleteAllSessions () : Observable<any> {
    const routeQuery=this.url+"/deleteAllSessions";
    return this.http.delete(routeQuery);
  }

  public addProjectToSession(sessionId : number,projectId : number) : Observable<number>
  {
    console.log("add project service : ",sessionId,projectId);
    let routeQuery=this.url+"/addProjectToSession/"+sessionId+"/"+projectId;
    return this.http.get<number>(routeQuery);
  }

  public deleteProjectFromSession(projectId : Number, sessionId : Number) : Observable<any>
  {
    const routeQuery=this.url+"/deleteProjectFromSession/"+sessionId+"/"+projectId;
    return this.http.delete(routeQuery);
  }

  public getSessionRuns(sessionId : number) : Observable<Array<Run>>
  {
    const routeQuery=this.url+"/getSessionRuns/"+sessionId;
    return this.http.get<Array<Run>>(routeQuery);
  }

  public getSessionsDepot() : Observable<Array<Session>>
  {
    const routeQuery=this.url+"/getSessionsDepot";
    return this.http.get<Array<Session>>(routeQuery);
  }

  public exportCSV(runId : number)
  {
    const headers = new HttpHeaders({Accept: 'text/csv'});
    let routeQuery=this.url+"/exportCSV/"+runId;
    return this.http.get(routeQuery,{headers,responseType:'arraybuffer'});
  }

}
